mcstatus/protocol/connection.py

```diff
@@ -20,199 +22,501 @@
 
 
 def ip_type(address: Union[int, str]) -> Optional[int]:
+    """Returns what version of IP a given address is."""
     try:
         return ip_address(address).version
     except ValueError:
         return None
 
 
-class Connection:
-    def __init__(self):
-        self.sent = bytearray()
-        self.received = bytearray()
+class BaseWriteSync(ABC):
+    """Base synchronous write class"""
 
-    def read(self, length: int) -> bytearray:
-        result = self.received[:length]
-        self.received = self.received[length:]
-        return result
+    __slots__ = ()
 
+    @abstractmethod
     def write(self, data: Union["Connection", str, bytearray, bytes]) -> None:
-        if isinstance(data, Connection):
-            data = data.flush()
-        if isinstance(data, str):
-            data = bytearray(data, "utf-8")
-        self.sent.extend(data)
+        """Write data to self."""
+        ...
 
-    def receive(self, data: Union[BytesConvertable, bytearray]) -> None:
-        if not isinstance(data, bytearray):
-            data = bytearray(data)
-        self.received.extend(data)
+    def __repr__(self) -> str:
+        """Return representation of self."""
+        return f"<{self.__class__.__name__} Object>" ""
 
-    def remaining(self) -> int:
-        return len(self.received)
+    @staticmethod
+    def _pack(format_: str, data: int) -> bytes:
+        """Pack data in with format in big-endian mode."""
+        return struct.pack(">" + format_, data)
 
-    def flush(self) -> bytearray:
-        result = self.sent
-        self.sent = bytearray()
-        return result
+    def write_varint(self, value: int) -> None:
+        """
+        Write varint with value value to self.
+
+        Max: 2 ** 31 - 1, Min: -(2 ** 31).
+        Raises ValueError if varint is out of range.
```

Comment:
We using RST format for docstrings, so you should use `:raises ValueError: ...`.
https://peps.python.org/pep-0287/#docstring-significant-features

```suggestion
        :raises ValueError: If varint is out of range.
```

Apllies to all non-RST docstring formatting.

Reply:
Changed in a09673e